### PR TITLE
Physics new code

### PR DESCRIPTION
**Pull request by avivfaraj**
### **Describe your change:**
Added new file to Physics folder. This file contain a function to calculate gamma function which is a common tool used in physics.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by Rohit-wed** [comment]
Wait for me do list is endless

File: physics/gamma_function.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        # Ensure input is a valid number
+        if number:
+            # Ensure input is an integer
+            if (isinstance(gamma(num), int)):
+                # Print result
+                print(gamma(num))
+            # Otherwise print results with √π (gamma of 0.5 is √π)
+            # Therefore all results will be a number times √π
+            else:
+                results = "{result:.4f}".format(result=gamma(num) / sqrt(pi))
```
[the remark this answers is not in the document]
Comment (avivfaraj) [reply physics/gamma_function.py]: Changed line 79 to f-string

**Review by QuantumNovice — CHANGES_REQUESTED** [review]

File: physics/gamma_function.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    """
+    if num == 1:
+        return 1
+    elif num == 0.5:
+        return sqrt(pi)
+    elif num > 1:
+        return (num - 1) * gamma(num - 1)
+    else:
+        # Error
+        return -2
```
[the remark this answers is not in the document]
Comment (avivfaraj) [reply physics/gamma_function.py]: Changed. 
Thanks

**Comment by QuantumNovice** [comment]
One last request. 
Can you rename the file to `gamma_recursive.py` and add it under `maths` directory?
We already have a gamma function based on quadrature. That seems like a suitable directory for it.

Good Work.👍 👍 👍

**Comment by avivfaraj** [comment]
> One last request.
> Can you rename the file to `gamma_recursive.py` and add it under `maths` directory?
> We already have a gamma function based on quadrature. That seems like a suitable directory for it.
> 
> Good Work.👍 👍 👍

Done.

Thank You.

**Comment by QuantumNovice** [comment]
@cclauss have a look at this.

**Comment by SayanMaity234** [comment]
It is working

File: maths/gamma_recursive.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+
+def gamma(num : Union[int, float]) -> Union[int, float]:
+    """
+    Calculates the value of Gamma function of num
+    where num is either an integer (1,2,3..) or a half-integer (0.5,1.5,2.5...).
+    Implemented using recursion
+    Examples:
+    >>> Gamma of: 0.5
+    √π
```
Comment (cclauss) [inline maths/gamma_recursive.py:20]: On your machine, please try `python3 -m doctest -v maths/gamma_recursive.py` to see if your doctests pass.

File: maths/gamma_recursive.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+from typing import Union
+
+
+def gamma(num : Union[int, float]) -> Union[int, float]:
+    """
+    Calculates the value of Gamma function of num
+    where num is either an integer (1,2,3..) or a half-integer (0.5,1.5,2.5...).
+    Implemented using recursion
+    Examples:
+    >>> Gamma of: 0.5
```
Comment (cclauss) [inline maths/gamma_recursive.py:19]:
```suggestion
    >>> gamma(0.5)
```

File: maths/gamma_recursive.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    """
+    assert sqrt(pi) == gamma(0.5)
+    assert 1 == gamma(1)
+    assert 1 == gamma(2)
+
+
+if __name__ == "__main__":
+    # Initialize boolean
+    number = True
+    # Get input from user
```
Comment (cclauss) [inline maths/gamma_recursive.py:48]:
```suggestion
    is_number = True
```
The comments do not say more than the code does so drop them.

`is_` makes it clear to the reader that this variable is a Boolean.

File: maths/gamma_recursive.py (hunk 3 above)
Comment (avivfaraj) [reply maths/gamma_recursive.py]: Done.

Thanks

File: maths/gamma_recursive.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    Calculates the value of Gamma function of num
+    where num is either an integer (1,2,3..) or a half-integer (0.5,1.5,2.5...).
+    Implemented using recursion
+    Examples:
+    >>> gamma(0.5)
+    1.7724538509055159
+    >>> gamma(2)
+    1
+    >>> gamma(3.5)
+    3.3233509704478426
```
Comment (cclauss) [inline maths/gamma_recursive.py:24]: We need some doctest that raise ValueErrors so that we not only see how the algorithm work but also how it fails.

Comment (avivfaraj) [reply maths/gamma_recursive.py:24]: Within the gamma function I didn't check the type of the argument. I did it in the main program before calling the function, so in the doctest I included negative numbers as arguments, but not strings. Let me know if that works or need to be modified.

Thanks

**Review by cclauss — APPROVED** [review]
Nice work.  I modified the exceptions to more closely mimic the Python's Standard Library [`math.gamma()`](https://docs.python.org/3/library/math.html#math.gamma) function.

**Review by Dummy59 — CHANGES_REQUESTED** [review]
> ~~cecf43d6481173e831af829da77911e1a3868a6c~~

**Comment by cclauss** [comment]
@Dummy59 Can you explain what was done here?

**Pull request merged**